File: src/main/scala/com/vc/medium/L654.scala
package com.vc.medium


class TreeNode654(var _value: Int) {
  var value: Int = _value
  var left: TreeNode654 = null
  var right: TreeNode654 = null
}

object L654 {
  def constructMaximumBinaryTree(nums: Array[Int]): TreeNode654 = {
    if(nums.length > 0) {
      val maxIdx = getMaxIdx(nums)
      val root = new TreeNode654(nums(maxIdx))
      root.left = constructMaximumBinaryTree(nums.slice(0, maxIdx))
      root.right = constructMaximumBinaryTree(nums.slice(maxIdx + 1, nums.length))
      return root
    }
    null
  }

  def getMaxIdx(nums: Array[Int]): Int = {
    var max = Int.MinValue
    var maxIdx = -1
    nums.indices.foreach(i => {
      if(nums(i) > max) {
        max = nums(i)
        maxIdx = i
      }
    })
    maxIdx
  }

}

File: src/main/java/com/vc/medium/L654.java
package com.vc.medium;

import com.vc.hard.TreeNode;

class L654 {
    private int[] arr;

    public TreeNode constructMaximumBinaryTree(int[] arr) {
        if(arr == null || arr.length == 0) return null;
        this.arr = arr;
        return helper(0, arr.length - 1);
    }

    private TreeNode helper(int from, int to) {
        if(from > to) return null;
        if(from == to) return new TreeNode(arr[from]);

        int maxIndex = getMaxIndex(from, to);
        TreeNode root = new TreeNode(arr[maxIndex]);
        root.left = helper(from, maxIndex - 1);
        root.right = helper(maxIndex + 1, to);
        return root;
    }

    private int getMaxIndex(int from, int to) {
        int maxIndex = from, maxValue = arr[from];
        for(int i = from + 1; i <= to; i++) {
            if(maxValue < arr[i]) {
                maxValue = arr[i];
                maxIndex = i;
            }
        }
        return maxIndex;
    }
}
